lib/internal/process/esm_loader.js

```diff
@@ -31,31 +40,93 @@ exports.importModuleDynamicallyCallback = async function(wrap, specifier) {
   throw new ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING();
 };
 
-let ESMLoader = new Loader();
-exports.ESMLoader = ESMLoader;
-
-exports.initializeLoader = initializeLoader;
-async function initializeLoader() {
-  const { getOptionValue } = require('internal/options');
-  const userLoader = getOptionValue('--experimental-loader');
-  if (!userLoader)
-    return;
-  let cwd;
-  try {
-    cwd = process.cwd() + '/';
-  } catch {
-    cwd = 'file:///';
+let processLoader;
+exports.getLoader = () => {
+  if (!processLoader) {
+    processLoader = (async () => {
+      const loader = new Loader();
+
+      const { getOptionValue } = require('internal/options');
+      const userLoaders = getOptionValue('--experimental-loader');
+
+      if (userLoaders.length > 0) {
+        const { emitExperimentalWarning } = require('internal/util');
+        emitExperimentalWarning('--experimental-loader');
+
+        const importedLoaders = [];
+        for (let i = 0; i < userLoaders.length; i += 1) {
+          const ns = await loader.importLoader(userLoaders[i]);
+
+          if (ns.resolve !== undefined) {
+            validateFunction(ns.resolve, 'resolve');
+          }
+          if (ns.getFormat !== undefined) {
+            validateFunction(ns.getFormat, 'getFormat');
+          }
+          if (ns.getSource !== undefined) {
+            validateFunction(ns.getSource, 'getSource');
+          }
+          if (ns.transformSource !== undefined) {
+            validateFunction(ns.transformSource, 'transformSource');
+            ArrayPrototypePush(loader.transformSourceHooks, ns.transformSource);
+          }
+
+          if (ns.getGlobalPreloadCode !== undefined) {
+            validateFunction(ns.getGlobalPreloadCode, 'getGlobalPreloadCode');
+            const code = ReflectApply(ns.getGlobalPreloadCode, undefined, []);
+            validateString(code, 'getGlobalPreloadCode return value');
+            ArrayPrototypePush(loader.globalPreloadCode, code);
+          }
+
+          importedLoaders.push(ns);
+        }
+
+        let cursor = ObjectFreeze({
+          __proto__: null,
+          resolve: loader._resolve,
+          getFormat: loader._getFormat,
+          getSource: loader._getSource,
+        });
+        for (let i = userLoaders.length - 1; i >= 0; i -= 1) {
```

Comment:
Let's get this left to right in, then I think we are almost there?

```suggestion
        for (let i = 0; i < userLoaders.length; i++) {
```

Reply:
It's already left to right. That would make it right to left.

Reply:
Ok, let me clearer:

```
node x.js --experimental-loader a --experimental-loader b
```

should call the resolve hook of `b` before it calls the resolve hook of `a`.

That is the layering happens from left to right, but the execution unwinds from right to left.

Reply:
In short, the last added loader takes ownership over from the one before.

Reply:
That is, left to right is about "authoritativeness" moving from left to right, not about call order. I can get behind reversing the order but then we should do the same for translate.

Reply:
it makes the first one the least authoritative because it only gets called if every other hook doesn't give a response. atm the first one is the most powerful because it can explicitly call to the rest of the chain and decide what to do with the result.

Reply:
> it makes the first one the least authoritative because it only gets called if every other hook doesn't give a response

Exactly, so without this change the first loader is the _most authoritative_. But because the transform hook chains, it is not authoritative over the final source that is executed, despite being authoritative over everything else. It is this "confusion of authority" that I specifically have issue with, and this is actually a critical block for me as authority must be clear.

Reply:
@guybedford i guess we can flip the order of the transform hooks